File: typescript/src/trinsicClientCredentials.ts
import { ServiceClientCredentials, WebResource, ServiceClientOptions } from "@azure/ms-rest-js";
import { CredentialsServiceClientOptions } from "./credentials/models"
import { WalletServiceClientOptions } from "./wallet/models"
import { ProviderServiceClientOptions } from "./provider/models"

export class TrinsicCredentialsClientOptions implements CredentialsServiceClientOptions {
    noRetryPolicy?: boolean;
    baseUri?: string;
}

export class TrinsicWalletClientOptions implements WalletServiceClientOptions {
    noRetryPolicy?: boolean
    baseUri?: string;
}

export class TrinsicProviderClientOptions implements ProviderServiceClientOptions {
    noRetryPolicy?: boolean
    baseUri?: string;
}

export class TrinsicClientCredentials implements ServiceClientCredentials {
    accessToken: string
    subscriptionKey: string;

    constructor(accessToken: string, subscriptionKey: string) {
        this.accessToken = accessToken;
        this.subscriptionKey = subscriptionKey;
    }

    signRequest(webResource: WebResource): Promise<WebResource> {
        webResource.headers.set("Authorization", `Bearer ${this.accessToken}`);
        webResource.headers.set("X-Trinsic-Subscription-Key", this.subscriptionKey);
        return Promise.resolve(webResource);
    }
}

export class TrinsicProviderCredentials implements ServiceClientCredentials {
    accountSecret: string

    constructor(accountSecret: string) {
        this.accountSecret = accountSecret;
    }

    signRequest(webResource: WebResource): Promise<WebResource> {
        webResource.headers.set("Authorization", `Bearer ${this.accountSecret}`);
        return Promise.resolve(webResource);
    }
}

File: csharp/TrinsicClientCredentials.cs
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.Extensions.Options;
using Microsoft.Rest;
// ReSharper disable CheckNamespace

namespace Microsoft.Extensions.DependencyInjection
{
    public class TrinsicClientCredentials : ServiceClientCredentials
    {
        /// <summary>
        /// Gets the options.
        /// </summary>
        /// <value>
        /// The options.
        /// </value>
        public ServiceClientOptions Options { get; }

        /// <summary>
        /// Initializes a new instance of the <see cref="TrinsicClientCredentials"/> class.
        /// </summary>
        /// <param name="options">The options.</param>
        public TrinsicClientCredentials(IOptions<ServiceClientOptions> options)
        {
            Options = options.Value;
        }

        /// <inheritdoc />
        public override Task ProcessHttpRequestAsync(HttpRequestMessage request, CancellationToken cancellationToken)
        {
            request.Headers.Add("Authorization", $"Bearer {Options.AccessToken}");
            request.Headers.Add("X-Trinsic-Subscription-Key", Options.SubscriptionKey);
            return base.ProcessHttpRequestAsync(request, cancellationToken);
        }
    }
}